### PR TITLE
Automatically generate networks.json

### PR DESCRIPTION
**Pull request by fedgiac**
Closes #26.

Inspired by [this comment](https://github.com/cowprotocol/ethflowcontract/pull/26#pullrequestreview-1113701155), it generates `networks.json` based on the content of our `broadcast` folder. When we manually deploy the contract we still need to commit the content of `broadcast`, however the file `networks.json` will be automatically generated along with the artifacts with our release script.

Unfortunately I don't think it's possible to generate `networks.json` in the deployment script: the main issue is that the transaction id isn't available inside the script. Otherwise, Foundry allows you to read from and write to files, as well as some basic JSON parsing (but not encoding).

### Test plan

Release a new tag under this commit and see the output.
[Here](https://github.com/cowprotocol/ethflowcontract/actions/runs/3347321333/jobs/5545176777#step:4:83) is the output of the GitHub action, [here](https://github.com/cowprotocol/ethflowcontract/commit/9dfdc62418670e49e574e1a2b19107ff75c75f48) is the new commit, and [here](https://github.com/cowprotocol/ethflowcontract/blob/9dfdc62418670e49e574e1a2b19107ff75c75f48/networks.json) is `networks.json`.

### CODE REVIEW

File: .github/scripts/release-artifacts.sh [hunk 1]
```diff
@@ -34,9 +35,9 @@ artifacts_tag="$target_ref_short-artifacts"
 git fetch origin "$target_ref"
 git checkout --detach "$target_ref"
 
-forge build -o artifacts
+bash "$script_path/generate-artifacts.sh"
 
-git add artifacts
+git add --all
 git commit -m "Add artifacts for $target_ref"
 git tag -m "Artifacts for $target_ref" --end-of-options "$artifacts_tag"
```
Comment (bh2smith) [inline .github/scripts/release-artifacts.sh:43]: I didn't realize we could commit to the repo from within GitHub actions! This is great news.

**Review by bh2smith — APPROVED** [review]
Some of this stuff is pretty fancy, but the description with all the references is plenty enough to approve. Nice stuff here!

File: .github/scripts/generate-artifacts.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+for deployment in "$deployment_output_folder/"*; do
+  # The subfolder name is the chain id
+  chain_id=${deployment##*/}
+  # First, every single deployment is formatted as if it had its own networks.json
+  jq --arg chainId "$chain_id" '
+    .transactions[]
+    | select(.transactionType == "CREATE" )
+    | {(.contractName): {($chainId): {address: .contractAddress, transactionHash: .hash }}}
+  '  <"$deployment/deployment.json"
+  # Then, all these single-contract single-chain-id networks.jsons are merged
```
Comment (josojo) [inline .github/scripts/generate-artifacts.sh:23]: Should this comment go one line further down, as it is done after the for loop is done?

File: .github/scripts/generate-artifacts.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+deployment_output_folder="$repo_root_dir/broadcast/Deploy.sol"
+
+echo "Creating build artifacts..."
+forge build -o artifacts
+
+echo "Creating networks.json..."
+for deployment in "$deployment_output_folder/"*; do
+  # The subfolder name is the chain id
+  chain_id=${deployment##*/}
+  # First, every single deployment is formatted as if it had its own networks.json
```
Comment (josojo) [inline .github/scripts/generate-artifacts.sh:17]: Could you make a comment on the assumptions about double deployments you are making? They are not allowed at all, right?

**Review by josojo — COMMENTED** [review]
Looks already good.

Why didn't you split the building of the artifacts and creating the networks json into two separate tasks/jobs?

**Comment by fedgiac** [comment]
> Why didn't you split the building of the artifacts and creating the networks json into two separate tasks/jobs?

Two reasons:

1. I want everything to be available at the head of `*-artifacts`. To do this by splitting it into two jobs, the two jobs would have to be run sequentially, the first would have to create a new commit, and the second would have to check out this commit and build a tag on top. Overall, it seemed too much overhead to me for a relatively simple task.
2. I like the idea of a release script that generates every automatically generated component that we want to make available. This way, for local testing we can just run the release script without being encumbered with the git stuff (or GitHub Action stuff).

File: .github/scripts/generate-artifacts.sh (hunk 2 above)
Comment (fedgiac) [reply .github/scripts/generate-artifacts.sh]: Fixed! Adding comments to inline commands in Bash is a bit tricky, which is why I cheated my way out by writing it one line before.

File: .github/scripts/generate-artifacts.sh (hunk 3 above)
Comment (fedgiac) [reply .github/scripts/generate-artifacts.sh:17]: Added a comment. I'd say the main problems with double deployments (i.e., deploying the same contract twice in the same run of the `Deploy.sol` script) is the `networks.json` format itself more than the execution of this script. Keys in a JSON files aren't supposed to repeat, so we'd have to find a different file format if we really want to capture information about two same-contract deployments. It is tricky however if for some reason we deploy two contracts of the same type but we only care about the first; then we need to change this script.

**Review by josojo — APPROVED** [review]

**Pull request merged**